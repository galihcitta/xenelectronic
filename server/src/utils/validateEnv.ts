import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
