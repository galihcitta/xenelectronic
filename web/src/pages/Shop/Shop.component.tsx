import React, { useState } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { Wrapper } from './Shop.styles';

import { FilterPanel } from './shop/FilterPanel.component';
import { ShopList } from './shop/ShopList.component';
import { GlobalSpinner } from '../../components/GlobalSpinner/GlobalSpinner.component';

import { Store } from '../../modules/rootReducer';
import { selectIsFetching } from '../../modules/shop/shop.selectors';

interface ShopSelection {
  isFetching: boolean;
}

export interface ShopProps extends ShopSelection {}

export const _Shop: React.FC<ShopProps> = ({ isFetching }) => {
  const [gridColumns, setGridColumns] = useState<number>(2);

  if (isFetching) return <GlobalSpinner />;

  return (
    <Wrapper>
      <FilterPanel gridColumns={gridColumns} setGridColumns={setGridColumns} />
      <ShopList gridColumns={gridColumns} />
    </Wrapper>
  );
};

const mapStateToProps = createStructuredSelector<Store, ShopSelection>({
  isFetching: selectIsFetching
});

export const Shop = connect(mapStateToProps)(_Shop);
