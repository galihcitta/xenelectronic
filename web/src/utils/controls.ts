import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
