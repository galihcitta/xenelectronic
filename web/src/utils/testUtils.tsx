import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
