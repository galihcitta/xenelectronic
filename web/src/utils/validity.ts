import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
