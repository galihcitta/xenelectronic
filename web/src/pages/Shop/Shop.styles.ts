import styled from 'styled-components';

import { device } from '../../utils/styles';

export const Wrapper = styled.section`
  width: 100%;
  display: flex;
  flex-direction: column;
  max-width: 500px;
  margin: 0 auto;
  margin-bottom: 1.4em;

  @media ${device.tablet} {
    max-width: 100%;
  }

  @media ${device.laptop} {
    padding-top: 1.2em;
  }

  button {
    max-width: 320px;
    align-self: center;
  }
`;
