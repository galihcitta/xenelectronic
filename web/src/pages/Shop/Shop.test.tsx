import { render } from '@testing-library/react';
import { withProvider } from '../../utils/testUtils';
import { _Shop, ShopProps } from './Shop.component';

const ShopWithProvider = withProvider(_Shop);

describe('<Shop />', () => {
  const props: ShopProps = {
    isFetching: false
  };

  const setup = (props: ShopProps) => {
    const utils = render(ShopWithProvider(props));
    return { ...utils };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders spinner when `isFetching` prop is true', () => {
    const { getByTestId } = setup({ isFetching: true });
    expect(getByTestId('Spinner')).toBeInTheDocument();
  });
});
