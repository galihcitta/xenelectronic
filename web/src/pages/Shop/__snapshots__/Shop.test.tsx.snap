// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Shop /> renders correctly 1`] = `
<DocumentFragment>
  <section
    class="sc-bdVaJa bmsbIv"
  >
    <div
      class="sc-bwzfXH fsMUGo"
    >
      <div
        class="sc-EHOje hSbnRA"
      >
        <label
          class="sc-bZQynM zXWjd"
          for="color"
        >
          Color
        </label>
        <select
          class="sc-gzVnrw jMwgLB"
          id="color"
          name="color"
        >
          <option
            class="sc-htoDjs dzhsTf"
            value="all"
          >
            all  (0)
          </option>
        </select>
      </div>
      <div
        class="sc-EHOje hSbnRA"
      >
        <label
          class="sc-bZQynM zXWjd"
          for="character"
        >
          Character
        </label>
        <select
          class="sc-gzVnrw jMwgLB"
          id="character"
          name="character"
        >
          <option
            class="sc-htoDjs dzhsTf"
            value="all"
          >
            all  (0)
          </option>
        </select>
      </div>
      <div
        class="sc-gZMcBi gocsoC"
      >
        <label
          class="sc-VigVT bFAbFJ"
          for="price"
        >
          Price
        </label>
        <input
          class="sc-gqjmRU Rvdhp"
          id="price"
          max="500"
          min="0"
          name="price"
          type="range"
          value="500"
        />
      </div>
      <div
        class="sc-htpNat eORBik"
      >
        <button
          class="sc-bxivhb kqNkey"
        >
          <i
            class="fas fa-th-large"
          />
        </button>
        <button
          class="sc-bxivhb cJMddQ"
        >
          <i
            class="fas fa-th"
          />
        </button>
      </div>
    </div>
    <p
      class="sc-jzJRlG eFBSZV"
    >
      No results.
    </p>
  </section>
</DocumentFragment>
`;
